url2b9f720f64_jfimbett_QBnB_tgz-pJ8-tests_TestJ8/src/model/Model.java

```diff
@@ -0,0 +1,988 @@
+	}
+	
+	/**
+	 * Returns the number of slack variables in the problem
+	 * @return num: The number of slack variables in the problem
+	 */
+	public int numberOfSlackVariables(){
+		int num=0;
+		for (int i = 0; i < constraints.size(); i++) {
+			//Equality constraint
+			if(constraints.get(i).getType()==0){
+				
+			}
+			else{
+				num++;
+			}
+		}
+		return num;
+	}
+
+
+	/**
+	 * Creates a bound by adding the constraint x<=bound or x>= bound
+	 * This method is used in Branch and Bounding for adding constraints to the initial problem
+	 * @param varname: Name of the variable to bound
+	 * @param value: Name of the bound
+	 * @param sign: Side of the restriction <= -> 1, >= -> 2
+	 * @return
+	 */
+	public double[][] addEqualityConstraintA(String varname, double value, boolean sign){
+		//<=
+		A=null;
+		if(sign){
+			QBnBVariable var= findVarByName(varname);
+			c=addToVector(0, c);
+		   
+			QBnBLinExp le= new QBnBLinExp();
+			le.addTerm(1, var);
+			QBnBconstr c= new QBnBconstr(le, 1, 1, "B"+varname+"<="+value, this);
+			this.addConst(c);
+			
+			update();
+		}
+		else{
+			QBnBVariable var= findVarByName(varname);
+			c=addToVector(0, c);
+		   
+			QBnBLinExp le= new QBnBLinExp();
+			le.addTerm(1, var);
+			QBnBconstr c= new QBnBconstr(le, 2, 1, "B"+varname+"<="+value, this);
+			this.addConst(c);
+			
+			update();
+		}
+	
+	return this.A;
+		
+	}
+	
+	/**
+	 * Adds value after the last position of a vector
+	 * @param val: Value to be added
+	 * @param vector: Original vector
+	 * @return newVector: Original vector plus the value added. 
+	 */
+	public double[] addToVector(double val, double[] vector){
+		double[] newVector=new double[vector.length+1];
+		for (int i = 0; i < newVector.length; i++) {
+			if(i<vector.length){
+				newVector[i]=vector[i];
+			}
+			else{
+				newVector[i]=val;
+			}
+		}
+		return newVector;
+	}
+	
+	/**
+	 * Returns the variable with name "name"
+	 * @param name, the name of the variable to find
+	 * @return the variable with name "name"
+	 */
+	public QBnBVariable findVarByName(String name){
+		
+		for (int i = 0; i < variables.size(); i++) {
+			QBnBVariable var= variables.get(i);
+			if(variables.get(i).getName().equals(name)) return var;
+		}
+		return null;
+	}
+	
+	/**
+	 * Returns the position of a variable using its name
+	 * @param name: The name of the variable
+	 * @return pos= Position of the variable
+	 */
+	public int findByName(String name){
+		int pos=0;
+		for (int i = 0; i < I.length; i++) {
+			String n=I[i];
+			if(name.equalsIgnoreCase(n)){
+				pos=i;
+				break;
+			}
+		}
+		return pos;
+	}
+	
+	/**
+	 * Indicates if the problem is a MIP
+	 * @return true if problem is MIP
+	 * false otherwise
+	 */
+	public boolean isInteger() {
+		boolean is=true;
+		double[] sol=simplex.getXb();
+		String[] names=simplex.getIb();
+		for (int i = 0; i < sol.length; i++) {
+			if(isInt(names[i])){
+				if(!doubleIsInteger(sol[i])){
+					is=false;
+					break;
+				}	
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 670:677, 681:688.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.